src/core/batch.rkt

```diff
@@ -82,71 +82,173 @@
+       (define half-idx (append-node 1/2))
+       (hash-set! mappings n (append-node `(* ,half-idx ,add-idx)))]
+      [(list 'sinh args)
+       (define exp-idx (append-node `(exp ,(hash-ref mappings args))))
+       (define one-idx (append-node 1))
+       (define inv-exp-idx (append-node `(/ ,one-idx ,exp-idx)))
+       (define neg-idx (append-node `(neg ,inv-exp-idx)))
+       (define add-idx (append-node `(+ ,exp-idx ,neg-idx)))
+       (define half-idx (append-node 1/2))
+       (hash-set! mappings n (append-node `(* ,half-idx ,add-idx)))]
+      [(list 'tanh args)
+       (define exp-idx (append-node `(exp ,(hash-ref mappings args))))
+       (define one-idx (append-node 1))
+       (define inv-exp-idx (append-node `(/ ,one-idx ,exp-idx)))
+       (define neg-idx (append-node `(neg ,inv-exp-idx)))
+       (define add-idx (append-node `(+ ,exp-idx ,inv-exp-idx)))
+       (define sub-idx (append-node `(+ ,exp-idx ,neg-idx)))
+       (hash-set! mappings n (append-node `(/ ,sub-idx ,add-idx)))]
+      [(list 'asinh args)
+       (define mult-idx (append-node `(* ,(hash-ref mappings args) ,(hash-ref mappings args))))
+       (define one-idx (append-node 1))
+       (define add-idx (append-node `(+ ,mult-idx ,one-idx)))
+       (define sqrt-idx (append-node `(sqrt ,add-idx)))
+       (define add2-idx (append-node `(+ ,(hash-ref mappings args) ,sqrt-idx)))
+       (hash-set! mappings n (append-node `(log ,add2-idx)))]
+      [(list 'acosh args)
+       (define mult-idx (append-node `(* ,(hash-ref mappings args) ,(hash-ref mappings args))))
+       (define -one-idx (append-node -1))
+       (define add-idx (append-node `(+ ,mult-idx ,-one-idx)))
+       (define sqrt-idx (append-node `(sqrt ,add-idx)))
+       (define add2-idx (append-node `(+ ,(hash-ref mappings args) ,sqrt-idx)))
+       (hash-set! mappings n (append-node `(log ,add2-idx)))]
+      [(list 'atanh args)
+       (define neg-idx (append-node `(neg ,(hash-ref mappings args))))
+       (define one-idx (append-node 1))
+       (define add-idx (append-node `(+ ,one-idx ,(hash-ref mappings args))))
+       (define sub-idx (append-node `(+ ,one-idx ,neg-idx)))
+       (define div-idx (append-node `(/ ,add-idx ,sub-idx)))
+       (define log-idx (append-node `(log ,div-idx)))
+       (define half-idx (append-node 1/2))
+       (hash-set! mappings n (append-node `(* ,half-idx ,log-idx)))]
+      [(list op args ...)
+       (hash-set! mappings n (append-node (cons op (map (curry hash-ref mappings) args))))]
+      [(approx spec impl) (hash-set! mappings n (append-node (approx spec (hash-ref mappings impl))))]
+      [_ (hash-set! mappings n (append-node node))]))
+
+  (define roots* (vector-map (curry hash-ref mappings) roots))
+  (define nodes* (list->vector (reverse icache)))
+
+  ; This may be too expensive to handle simple 1/2, 1/3 and 2/3 zombie nodes..
+  #;(remove-zombie-nodes (batch nodes* roots* vars (vector-length nodes*)))
```

Comment:
I guess I don't really care but is it actually slow?

Reply:
I assumed that do a taylor expansion for `1/2` or `1/3` is gonna be faster than reconstructing the whole batch again and remove them.
I think we may use this function somewhere else, but here it seems to be okay. However, zombie nodes indeed can be introduced by `expand-taylor`

Reply:
Yeah, OK, agreed. Wait—why are these nodes ever zombies? Don't we only insert them if we use them?

Reply:
Oh, yeah, at the time we insert `1/2` we do not know that it is actually a part of `pow` that is to be erased since we are traversing "bottom-down" (or left-to-right when imaging nodes as a vector). And after we have already inserted `1/2` and realized that it is a part of `pow` - it is too late to delete because some other nodes may refer to `1/2`, `remove-zombie-nodes` has to be called.